### PR TITLE
ticktick: init at 1.0.80

### PR DESCRIPTION
**Pull request by hbjydev**
## Description of changes

This is a proprietary task management & time tracking app I use on iOS. It's a repackage of the public-available `.deb` packages they provide for Debian derivatives.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by nagy — CHANGES_REQUESTED** [review]
Can you follow the guidelines at https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions ? also the part in the maintainers list needs to be a separate commit. you can find good examples for the commit message in the history of that file.

File: pkgs/applications/office/ticktick/default.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
+{ lib, fetchurl, stdenv, wrapGAppsHook, dpkg, autoPatchelfHook,
+  glibc, gcc-unwrapped, nss, libdrm, mesa, alsa-lib, xdg-utils, systemd }:
```
Comment (TheBrainScrambler) [inline pkgs/applications/office/ticktick/default.nix:2]: Please expand these, so that there is one dependency per line

File: pkgs/applications/office/ticktick/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+let
+  pname = "ticktick";
+  version = "1.0.80";
+
+  src =
+    if stdenv.hostPlatform.system == "x86_64-linux" then
```
Comment (TheBrainScrambler) [inline pkgs/applications/office/ticktick/default.nix:8]: Isn't this already taken care by meta.platforms ?

File: pkgs/applications/office/ticktick/default.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
+{ lib, fetchurl, stdenv, wrapGAppsHook, dpkg, autoPatchelfHook,
+  glibc, gcc-unwrapped, nss, libdrm, mesa, alsa-lib, xdg-utils, systemd }:
+let
```
Comment (TheBrainScrambler) [inline pkgs/applications/office/ticktick/default.nix:3]: A better pattern would be to define `pname`, `version` and `src` directly in `stdenv.mkDerivation`, and to instead give it a function with an argument `finalAttrs` as argument, and then refer to `finalAttrs.pname` for example.

See https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes

**Review by TheBrainScrambler — CHANGES_REQUESTED** [review]

**Comment by TheBrainScrambler** [comment]
Ah and please also rename the title of this PR to "ticktick: init at 1.0.80" (include the version too)

File: pkgs/applications/office/ticktick/default.nix (hunk 2 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:8]: Was largely following `skypeforlinux`'s approach as an example of doing it with a DEB package, happy to remove it if it's not necessary.

File: pkgs/applications/office/ticktick/default.nix (hunk 3 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:3]: That doubles the length of a lot of my substitutions in the install phase, I feel it just makes it read a lot less cleanly.

**Comment by hbjydev** [comment]
> Ah and please also rename the title of this PR to "ticktick: init at 1.0.80" (include the version too)

Done

**Comment by hbjydev** [comment]
> also the part in the maintainers list needs to be a separate commit.

Is this going to need me to rewrite the whole PR? I'd need to go back to when I implemented that and redo the commit.

File: pkgs/applications/office/ticktick/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    sed -i \
+      "s|Exec=/opt/TickTick/ticktick|Exec=$out/bin/${final.pname}|" \
+      "$out/share/applications/${final.pname}.desktop"
+  '';
+
+  meta = with lib; {
+    description = ''
+      TickTick is a powerful to-do & task management app with seamless cloud
+      synchronization across all your devices.
+    '';
```
Comment (nagy) [inline pkgs/applications/office/ticktick/default.nix:67]: I would rather keep the description to one line and you dont need to use the name of it itself in the description. There is also `longDescription` if you want to have a place for extra text.

File: pkgs/applications/office/ticktick/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    homepage = "https://ticktick.com/home/";
+    license = licenses.unfree;
+    maintainers = with maintainers; [ hbjydev ];
+    platforms = [ "x86_64-linux" ];
```
Comment (nagy) [inline pkgs/applications/office/ticktick/default.nix:71]: Can you add `sourceProvenance = with sourceTypes; [ binaryNativeCode ];` ?

File: pkgs/applications/office/ticktick/default.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    systemd
+  ];
+
+  unpackPhase = ''
+    mkdir -p "$out/share" "$out/opt/${final.pname}" "$out/bin"
+    dpkg-deb --fsys-tarfile "$src" | tar --extract --directory="$out"
+  '';
+
+  installPhase = ''
+    cp -av $out/opt/TickTick/* $out/opt/${final.pname}
```
Comment (nagy) [inline pkgs/applications/office/ticktick/default.nix:54]: Please add `runHook preInstall` here and `runHook postInstall` at the end.

File: pkgs/applications/office/ticktick/default.nix [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    xdg-utils
+  ];
+
+  # Needed to make the process get past zygote_linux fork()'ing
+  runtimeDependencies = [
+    systemd
+  ];
+
+  unpackPhase = ''
+    mkdir -p "$out/share" "$out/opt/${final.pname}" "$out/bin"
```
Comment (nagy) [inline pkgs/applications/office/ticktick/default.nix:49]: Please add `runHook preUnpack` here and `runHook postUnpack` at the end.

**Review by nagy — CHANGES_REQUESTED** [review]
Also I dont think you need a new commit for the conversion to `finalAttrs`.

**Comment by hbjydev** [comment]
> Also I dont think you need a new commit for the conversion to finalAttrs.

It's a refactor I made after rebasing, so it still gets a commit

**Comment by nagy** [comment]
> It's a refactor I made after rebasing, so it still gets a commit

Sure, but you dont need a handful of commits if you just introduce a package, you can squash them. After that it looks good to me.

File: pkgs/applications/office/ticktick/default.nix [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    ln -sf "$out/opt/${final.pname}/${final.pname}" "$out/bin/${final.pname}"
+    sed -i \
+      "s|Exec=/opt/TickTick/ticktick|Exec=$out/bin/${final.pname}|" \
+      "$out/share/applications/${final.pname}.desktop"
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A powerful to-do & task management app with seamless cloud synchronization across all your devices.";
```
Comment (eclairevoyant) [inline pkgs/applications/office/ticktick/default.nix:72]:
```suggestion
    description = "A powerful to-do & task management app with seamless cloud synchronization across all your devices";
```

File: pkgs/applications/office/ticktick/default.nix [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+, xdg-utils
+, systemd
+}:
+stdenv.mkDerivation (final: {
+  pname = "ticktick";
+  version = "1.0.80";
+
+  src = fetchurl {
+    url = "https://d2atcrkye2ik4e.cloudfront.net/download/linux/linux_deb_x64/${final.pname}-${final.version}-amd64.deb";
+    sha256 = "1iyxn9rdv6p8d36swqpxb346arblk00yhv9z2qf6i6r2a4sbrbqh";
```
Comment (eclairevoyant) [inline pkgs/applications/office/ticktick/default.nix:22]: please use SRI hash format (`hash = "..."` instead of `sha256 = "..."`)

File: pkgs/applications/office/ticktick/default.nix [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  };
+
+  system = "x86_64-linux";
```
Comment (eclairevoyant) [inline pkgs/applications/office/ticktick/default.nix:25]: why is this here?

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
for a _new_ package like this, should all be one commit for adding the package. therefore 2 commits total (one for adding to all-maintainers, one for the package itself)

File: pkgs/applications/office/ticktick/default.nix [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+, glibc
+, gcc-unwrapped
+, nss
+, libdrm
+, mesa
+, alsa-lib
+, xdg-utils
+, systemd
+}:
+stdenv.mkDerivation (final: {
```
Comment (TheBrainScrambler) [inline pkgs/applications/office/ticktick/default.nix:16]: The convention is to name the argument finalAttrs, could you change that ?

File: pkgs/applications/office/ticktick/default.nix [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    cp -av $out/opt/TickTick/* $out/opt/${final.pname}
+    cp -av $out/usr/share/* $out/share
+    rm -rf $out/usr $out/opt/TickTick
+    ln -sf "$out/opt/${final.pname}/${final.pname}" "$out/bin/${final.pname}"
+    sed -i \
```
Comment (TheBrainScrambler) [inline pkgs/applications/office/ticktick/default.nix:64]: I've just realised you could use `substituteInPlace` instead of `sed` here.
https://nixos.org/manual/nixpkgs/stable/#fun-substitute
```
substituteInPlace "$out/share/applications/${final.pname}.desktop" \
  --replace "Exec=/opt/TickTick/ticktick" "Exec=$out/bin/${final.pname}"
```
(didn't test this code)

**Review by TheBrainScrambler — CHANGES_REQUESTED** [review]

**Comment by hbjydev** [comment]
> > It's a refactor I made after rebasing, so it still gets a commit
> 
> Sure, but you dont need a handful of commits if you just introduce a package, you can squash them. After that it looks good to me.

I'll squash once I finish work later then.

File: pkgs/applications/office/ticktick/default.nix (hunk 12 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:64]: @TheBrainScrambler What's the motivation for using `substituteInPlace` over `sed`? For future reference mainly, that `sed` should work on every build.

File: pkgs/applications/office/ticktick/default.nix (hunk 11 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:16]: @TheBrainScrambler To shorten how much extra text was in the strings below, as I was told to convert from a `let ... in` setup to the function method here. I didn't like that I was adding 11 characters to _every_ instance of `${pname}`, it made the lines unnecessarily long. Plus, I'm a new packager and was unaware there was a convention here

File: pkgs/applications/office/ticktick/default.nix (hunk 10 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:25]: @eclairevoyant I mentioned earlier in the PR that I was following `skypeforlinux`'s example as another DEB repackage effort.

File: pkgs/applications/office/ticktick/default.nix (hunk 9 above)
Comment (hbjydev) [reply pkgs/applications/office/ticktick/default.nix:22]: Would that be `hash = "sha256:1iyxn9rdv6p8d36swqpxb346arblk00yhv9z2qf6i6r2a4sbrbqh"`?

File: pkgs/applications/office/ticktick/default.nix (hunk 10 above)
Comment (TheBrainScrambler) [reply pkgs/applications/office/ticktick/default.nix:25]: I would say you can remove it then if there's no reason to have it, after all we have meta.platforms. Not every package is a good example sadly

File: pkgs/applications/office/ticktick/default.nix (hunk 11 above)
Comment (TheBrainScrambler) [reply pkgs/applications/office/ticktick/default.nix:16]: Yes it's fine if you're unaware about the convention, especially that this one is fairly new, that's why reviews exist ;)

But I don't think shortening text is a good reason. Every package that I've seen using this pattern always named the argument finalAttrs. Consistency is a good thing since if you read someone else's package and see "finalAttrs" you immediately know what it is. If each people instead used their own names, then it will each time be a different one and it makes code hard to read.

File: pkgs/applications/office/ticktick/default.nix (hunk 12 above)
Comment (TheBrainScrambler) [reply pkgs/applications/office/ticktick/default.nix:64]: It's not strictly needed, but is more idiomatic and thus easier to read.
https://discourse.nixos.org/t/when-to-use-substituteinplace-functions-vs-a-patch/11073/10

**Review by TheBrainScrambler — APPROVED** [review]

**Review by eclairevoyant — APPROVED** [review]
